### PR TITLE
MsBuildWorkspace small fixes

### PR DESCRIPTION
**Pull request by jmarolf**

### CODE REVIEW

File: src/Workspaces/Core/MSBuild/MSBuild/Build/ProjectBuildManager.cs [hunk 1]
```diff
@@ -65,13 +61,12 @@ internal class ProjectBuildManager
...
         private MSBuildDiagnosticLogger? _batchBuildLogger;
-        private bool _batchBuildStarted;
 
         ~ProjectBuildManager()
         {
-            if (_batchBuildStarted)
+            if (BatchBuildStarted)
             {
-                new InvalidOperationException("ProjectBuilderManager.Stop() not called.");
+                throw new InvalidOperationException("ProjectBuilderManager.Stop() not called.");
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/MSBuild/MSBuild/Build/ProjectBuildManager.cs:69]: lol.

File: src/Workspaces/Core/MSBuild/MSBuild/DiagnosticReporter.cs [hunk 2]
```diff
@@ -2,31 +2,31 @@
...
+using System.ComponentModel;
 using Microsoft.CodeAnalysis.MSBuild.Logging;
+using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.MSBuild
 {
     internal sealed class DiagnosticReporter
     {
-        internal ImmutableList<WorkspaceDiagnostic> Diagnostics;
+        internal ImmutableList<WorkspaceDiagnostic> _diagnostics;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/MSBuild/MSBuild/DiagnosticReporter.cs:15]: the name before was correct.  it is exposed outside the type, so should be pascal cased.  `_` is for privates.

File: src/Workspaces/Core/MSBuild/MSBuild/MSBuildProjectLoader.SolutionFilterReader.cs [hunk 3]
```diff
@@ -18,16 +20,22 @@ public static bool IsSolutionFilterFilename(string filename)
...
                     var solutionPath = solution.GetProperty("path").GetString()?.Replace('\\', Path.DirectorySeparatorChar);
+                    if (solutionPath is null || Path.GetDirectoryName(filterFilename) is not string baseDirectory)
+                    {
+                        solutionFilename = string.Empty;
+                        projectFilter = ImmutableHashSet<string>.Empty;
+                        return false;
+                    }
 
-                    if (!pathResolver.TryGetAbsoluteSolutionPath(solutionPath, baseDirectory: Path.GetDirectoryName(filterFilename), DiagnosticReportingMode.Throw, out solutionFilename))
+                    if (!pathResolver.TryGetAbsoluteSolutionPath(solutionPath, baseDirectory, DiagnosticReportingMode.Throw, out solutionFilename))
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/MSBuild/MSBuild/MSBuildProjectLoader.SolutionFilterReader.cs:38]: this scares me.  skipping for now.

File: src/Workspaces/Core/MSBuild/MSBuild/MSBuildProjectLoader.Worker_ResolveReferences.cs [hunk 4]
```diff
@@ -349,8 +343,15 @@ private async Task<bool> VerifyProjectOutputExistsAsync(string projectPath, Reso
 
                 foreach (var projectFileInfo in projectFileInfos)
                 {
+                    if (projectFileInfo.IsEmpty)
+                    {
+                        continue;
+                    }
+
                     var outputFilePath = projectFileInfo.OutputFilePath;
+                    RoslynDebug.AssertNotNull(outputFilePath);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/MSBuild/MSBuild/MSBuildProjectLoader.Worker_ResolveReferences.cs:352]: Contract.ThrowIfNull?

File: src/Workspaces/Core/MSBuild/MSBuild/MSBuildWorkspace.cs [hunk 5]
```diff
@@ -104,7 +102,7 @@ public static MSBuildWorkspace Create(IDictionary<string, string> properties, Ho
         /// <summary>
         /// Diagnostics logged while opening solutions, projects and documents.
         /// </summary>
-        public ImmutableList<WorkspaceDiagnostic> Diagnostics => _reporter.Diagnostics;
+        public ImmutableList<WorkspaceDiagnostic> Diagnostics => _reporter._diagnostics;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/MSBuild/MSBuild/MSBuildWorkspace.cs:105]: i triply hate ths now.  having another type access an `_` member of another varaible looks like it's somehow accessing a privat.e

File: src/Workspaces/Core/MSBuild/MSBuild/MSBuildWorkspace.cs [hunk 6]
```diff
@@ -260,29 +258,29 @@ private static string GetAbsolutePath(string path, string baseDirectoryPath)
...
+            return projectResult;
         }
 
         #endregion
 
         #region Apply Changes
         public override bool CanApplyChange(ApplyChangesKind feature)
         {
-            switch (feature)
+            return feature switch
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/MSBuild/MSBuild/MSBuildWorkspace.cs:271]: honestly i dont' get why this is a switch, and not just `return X || Y || Z;`

**Comment by CyrusNajmabadi** [comment]
So i ended up skpping this.  lots of mechanical changes i approve of, and would accept a PR on.  I don't like intermixing it with things taht are sublte and could change semantics. c an you break apart?

**Comment by sharwell** [comment]
Marked as draft since it builds on #52389

**Pull request closed without merging**